In [72]:
!python -m textblob.download_corpora
!python -m nltk.downloader punkt_tab

[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Unzipping corpora/brown.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package conll2000 to /root/nltk_data...
[nltk_data]   Unzipping corpora/conll2000.zip.
[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Unzipping corpora/movie_reviews.zip.
Finished.
<frozen runpy>:128: RuntimeWarning: 'nltk.downloader' found in sys.modules after import of package 'nltk', but prior to execution of 'nltk.downloader'; this may result in unpredictable behaviour
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


1min para que corra

In [2]:
!pip install PyPDF2
!pip install pymupdf
!pip install flair
!pip install stanza
!pip install pyldavis
!pip install pdfplumber

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 232.6/232.6 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.0/20.0 MB 22.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 17.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 48.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.2/139.2 kB 11.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.8/44.8 kB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 202.6/202.6 kB 16.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.3/13.3 MB 86.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.2/84.2 kB 6.7 MB/s eta 0:00:00
  Created wheel for langdetect: filena

In [3]:
import stanza
stanza.download('en')

INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Downloading default packages for language: en (English) ...


INFO:stanza:Downloaded file to /root/stanza_resources/en/default.zip
INFO:stanza:Finished downloading models and saved to /root/stanza_resources


In [93]:
# Core NLP Libraries
import spacy  # For advanced NLP tasks
import nltk  # For traditional NLP tasks
from textblob import TextBlob  # For sentiment and basic analysis
from transformers import pipeline  # For advanced pre-trained models

# Text Preprocessing Libraries
import re  # Regular expressions for text cleaning
from gensim.models import Word2Vec  # For word embeddings and topic modeling

# Visualization Libraries
import matplotlib.pyplot as plt  # For plotting graphs
from wordcloud import WordCloud  # For generating word clouds
import seaborn as sns  # For advanced visualizations

# Sentiment Analysis Libraries
from nltk.sentiment.vader import SentimentIntensityAnalyzer  # For VADER sentiment analysis

# Topic Modeling and Clustering
from sklearn.feature_extraction.text import TfidfVectorizer  # For TF-IDF vectorization
from sklearn.decomposition import PCA  # For dimensionality reduction
from sklearn.cluster import KMeans  # For clustering
from sklearn.metrics.pairwise import cosine_similarity

# Text Embeddings
from sentence_transformers import SentenceTransformer  # For sentence embeddings

# Document Management
from PyPDF2 import PdfReader  # For extracting text from PDFs
import fitz  # (PyMuPDF) For advanced PDF handling
import pandas as pd  # For tabular data manipulation

# Comprehensive NLP Pipelines
import flair  # For entity recognition and sentiment analysis

import os
from nltk.corpus import stopwords
from google.colab import sheets

# Initialize nltk downloads (if necessary)
nltk.download('punkt')  # For tokenization
nltk.download('stopwords')  # For stopwords
nltk.download('vader_lexicon')  # For VADER sentiment analysis

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [5]:
os.chdir(r'/content/sample_data')

In [118]:
import pdfplumber

path = '/content/sample_data/FOMCpresconf20241218.pdf'

# Use pdfplumber to extract text
with pdfplumber.open(path) as pdf:
    text = " ".join([page.extract_text() for page in pdf.pages])



In [95]:
text

'December 18, 2024 Chair Powell’s Press Conference FINAL\nTranscript of Chair Powell’s Press Conference\nDecember 18, 2024\nCHAIR POWELL. Good afternoon. My colleagues and I remain squarely focused on\nachieving our dual-mandate goals of maximum employment and stable prices for the benefit of\nthe American people. The economy is strong overall and has made significant progress toward\nour, our goals over the past two years. The labor market has cooled from its formerly overheated\nstate and remains solid. Inflation has moved much closer to our 2 percent longer-run goal.\nWe’re committed to maintaining our economy’s strength by supporting maximum\nemployment and returning inflation to our 2 percent goal. To that end, today, the Federal Open\nMarket Committee decided to take another step in reducing the degree of policy restraint by\nlowering our policy interest rate by ¼ percentage point. We also decided to continue to reduce\nour securities holdings. I’ll have more to say about monetar

In [119]:
# Define patterns to remove unwanted elements
date_pattern = r"\b(January|February|March|April|May|June|July|August|September|October|November|December) \d{1,2} , \d{4}\b"  # Matches date format
title_pattern = r"Chair Powell ’s Press Conference"  # Matches title
version_pattern = r"FINAL"  # Matches transcript version
page_counter_pattern = r"Page \d+ of \d+"  # Matches page counter (e.g., "Page 1 of 10")

# Combine all patterns into one regex
unwanted_pattern = re.compile(
    f"({date_pattern}|{title_pattern}|{version_pattern}|{page_counter_pattern})"
)

# Clean the text by removing unwanted elements
cleaned_text = re.sub(unwanted_pattern, "", text)

# Function to split the text into Powell's remarks and journalist interventions
def split_press_conference(text, moderator_name="MICHELLE SMITH."):
    # Split by moderator's name to find where journalist interventions begin
    sections = re.split(fr"({moderator_name})", text)
    powell_remarks = sections[0].strip()  # Powell's opening remarks
    interventions = sections[1:]  # Everything after the moderator's first mention

    # Group moderator and journalist interventions
    split_data = []
    for i in range(1, len(interventions), 2):
        moderator = interventions[i - 1].strip()
        journalist_content = interventions[i].strip()
        split_data.append({"moderator": moderator, "content": journalist_content})
    return powell_remarks, split_data

# Split the press conference
powell_remarks, interventions = split_press_conference(cleaned_text)

# Create DataFrames for Powell's remarks and interventions
powell_remarks_df = pd.DataFrame({"content": [powell_remarks]})
interventions_df = pd.DataFrame(interventions)

# Save datasets
powell_remarks_path = "powell_opening_remarks.csv"
interventions_path = "journalist_interventions.csv"
powell_remarks_df.to_csv(powell_remarks_path, index=False)
interventions_df.to_csv(interventions_path, index=False)

print(f"Powell's opening remarks saved at: {powell_remarks_path}")
print(f"Journalist interventions saved at: {interventions_path}")

Powell's opening remarks saved at: powell_opening_remarks.csv
Journalist interventions saved at: journalist_interventions.csv


In [98]:
powell_remarks_df

content
0  December 18, 2024 Chair Powell’s Press Confere...

In [97]:
interventions_df

moderator                                            content
0   MICHELLE SMITH.  Jeanna.\nJEANNA SMIALEK. Jeanna Smialek with t...
1   MICHELLE SMITH.  Howard.\nHOWARD SCHNEIDER. Howard Schneider wi...
2   MICHELLE SMITH.  Chris.\nCHRIS RUGABER. Thank you. Chris Rugabe...
3   MICHELLE SMITH.  Nick.\nNICK TIMIRAOS. Nick Timiraos, the Wall ...
4   MICHELLE SMITH.  Mike.\nMICHAEL MCKEE. Michael McKee from Bloom...
5   MICHELLE SMITH.  Colby.\nCOLBY SMITH. Thank you. Colby Smith wi...
6   MICHELLE SMITH.  Steve.\nSTEVE LIESMAN. Mr. Chairman, I, I did ...
7   MICHELLE SMITH.  Neil.\nNEIL IRWIN. Thanks Chair, Chair Powell....
8   MICHELLE SMITH.  Andrew.\nANDREW ACKERMAN. Hey, happy holidays,...
9   MICHELLE SMITH.  Amara.\nAMARA OMEOKWE. Thank you. Amara Omeokw...
10  MICHELLE SMITH.  Elizabeth.\nELIZABETH SCHULZE. Thanks so much,...
11  MICHELLE SMITH.  Edward.\nEDWARD LAWRENCE. Thanks, Chair Powell...
12  MICHELLE SMITH.  Victoria.\nVICTORIA GUIDA. Hi, Victoria Guida ...
13  MICHELLE SMITH.  Kelly.\nKELLY O’GRADY. Thanks, Chair Powell. K...
14  MICHELLE SMITH.  Go to Nancy for the last question.\nNANCY MARS...

#Interventions_df

Delete moderator --- Split content: Journalist, Media, Question/Questions, Powell's answer

--Identify Journalist name, put it into the "Journalist" column

--Identify media, put it into the "Media" column

In [120]:
# Delete the 'moderator' column
if 'moderator' in interventions_df.columns:
    interventions_df = interventions_df.drop('moderator', axis=1)

# Now interventions_df no longer has the 'moderator' column

# Display the updated DataFrame (optional)
interventions_df

content
0   Jeanna.\nJEANNA SMIALEK. Jeanna Smialek with t...
1   Howard.\nHOWARD SCHNEIDER. Howard Schneider wi...
2   Chris.\nCHRIS RUGABER. Thank you. Chris Rugabe...
3   Nick.\nNICK TIMIRAOS. Nick Timiraos, the Wall ...
4   Mike.\nMICHAEL MCKEE. Michael McKee from Bloom...
5   Colby.\nCOLBY SMITH. Thank you. Colby Smith wi...
6   Steve.\nSTEVE LIESMAN. Mr. Chairman, I, I did ...
7   Neil.\nNEIL IRWIN. Thanks Chair, Chair Powell....
8   Andrew.\nANDREW ACKERMAN. Hey, happy holidays,...
9   Amara.\nAMARA OMEOKWE. Thank you. Amara Omeokw...
10  Elizabeth.\nELIZABETH SCHULZE. Thanks so much,...
11  Edward.\nEDWARD LAWRENCE. Thanks, Chair Powell...
12  Victoria.\nVICTORIA GUIDA. Hi, Victoria Guida ...
13  Kelly.\nKELLY O’GRADY. Thanks, Chair Powell. K...
14  Go to Nancy for the last question.\nNANCY MARS...

In [121]:
media_list = [
    "CNN Business",
    "Politico",
    "The Washington Post",
    "Bloomberg Radio and Television",
    "Bloomberg News",
    "Associated Press",
    "Reuters",
    "Marketplace",
    "American Banker",
    "The Economist",
    "Yahoo Finance",
    "Wall Street Journal",
    "CNBC",
    "MarketWatch",
    "Financial Times",
    "New York Times",
    "CBS News",
    "Punchbowl News",
    "Axios",
    "Fox Business",
    "NPR",
    "ABC",
    "Jefferies Financial Group",
    "HuffPost"
]


# Function to find the media name in a given text
def find_media(text):
    for media in media_list:
        if media.lower() in text.lower():  # Case-insensitive matching
            return media
    return None  # Return None if no media name is found

# Apply the function to the 'content' column and create a new column for media names
interventions_df["media"] = interventions_df["content"].apply(find_media)

# Display the updated DataFrame
print(interventions_df)

                                              content  \
0   Jeanna.\nJEANNA SMIALEK. Jeanna Smialek with t...   
1   Howard.\nHOWARD SCHNEIDER. Howard Schneider wi...   
2   Chris.\nCHRIS RUGABER. Thank you. Chris Rugabe...   
3   Nick.\nNICK TIMIRAOS. Nick Timiraos, the Wall ...   
4   Mike.\nMICHAEL MCKEE. Michael McKee from Bloom...   
5   Colby.\nCOLBY SMITH. Thank you. Colby Smith wi...   
6   Steve.\nSTEVE LIESMAN. Mr. Chairman, I, I did ...   
7   Neil.\nNEIL IRWIN. Thanks Chair, Chair Powell....   
8   Andrew.\nANDREW ACKERMAN. Hey, happy holidays,...   
9   Amara.\nAMARA OMEOKWE. Thank you. Amara Omeokw...   
10  Elizabeth.\nELIZABETH SCHULZE. Thanks so much,...   
11  Edward.\nEDWARD LAWRENCE. Thanks, Chair Powell...   
12  Victoria.\nVICTORIA GUIDA. Hi, Victoria Guida ...   
13  Kelly.\nKELLY O’GRADY. Thanks, Chair Powell. K...   
14  Go to Nancy for the last question.\nNANCY MARS...   

                             media  
0                   New York Times  
1            

In [122]:
# Function to clean text in the "content" column
def clean_text(text):
    # Remove line breaks and replace them with spaces
    text = text.replace("\n", " ")

    # Generalized regex for human sounds and pauses
    human_sounds_regex = r"\b(?:[Mm]+[Hh]\s[Hh]|[Uu]+[Mm]|[Hh]+[Mm]+|[Aa]+[Hh]|[Oo]+[Hh]|[Ee]+[Rr]*|[Uu]+[Hh]+|[Mm]+|[Hh]+|[Uu]+|[Pp]+[Hh]+|[Er]+)\b"

    # Remove human sounds using the regex
    text = re.sub(human_sounds_regex, "", text)

    # Fix stuttering (e.g., "s-sample" -> "sample")
    text = re.sub(r"\b([a-zA-Z])-\1\b", r"\1", text)

    # Eliminar prefijos 'b' si existen y no están seguidos de ninguna letra
    text = re.sub(r'\b(b)\b', '', text)

    # Keep numbers and periods while removing other special characters
    text = re.sub(r"[^a-zA-Z0-9.\sáéíóúüñçÁÉÍÓÚÜÑÇ]", '', text)

    # Replace NaN values with an empty string
    if pd.isnull(text):
        text = ""

    # Eliminar palabras vacías (stopwords)
    stop_words = set(stopwords.words('english'))  # Usar la lista de stopwords en inglés
    text = ' '.join(word for word in text.split() if word not in stop_words)

    # Remove all hyphens and replace them with a single space
    text = re.sub(r"-", " ", text)

    # Remove all apostrophes
    text = re.sub(r"'", "", text)

    # Replace multiple spaces with a single space
    text = re.sub(r"\s+", " ", text).strip()

    return text

# Apply the cleaning function to each cell in the "content" column
interventions_df["content"] = interventions_df["content"].apply(clean_text)

interventions_df

content  \
0   Jeanna. JEANNA SMIALEK. Jeanna Smialek New Yor...   
1   Howard. HOWARD SCHNEIDER. Howard Schneider Reu...   
2   Chris. CHRIS RUGABER. Thank you. Chris Rugaber...   
3   Nick. NICK TIMIRAOS. Nick Timiraos Wall Street...   
4   Mike. MICHAEL MCKEE. Michael McKee Bloomberg R...   
5   Colby. COLBY SMITH. Thank you. Colby Smith Fin...   
6   Steve. STEVE LIESMAN. Mr. Chairman I I hear us...   
7   Neil. NEIL IRWIN. Thanks Chair Chair Powell. N...   
8   Andrew. ANDREW ACKERMAN. Hey happy holidays Mr...   
9   Amara. AMARA OMEOKWE. Thank you. Amara Omeokwe...   
10  Elizabeth. ELIZABETH SCHULZE. Thanks much Chai...   
11  Edward. EDWARD LAWRENCE. Thanks Chair Powell q...   
12  Victoria. VICTORIA GUIDA. Hi Victoria Guida Po...   
13  Kelly. KELLY OGRADY. Thanks Chair Powell. Kell...   
14  Go Nancy last question. NANCY MARSHALLGENZER. ...   

                             media  
0                   New York Times  
1                          Reuters  
2                 Associated Press  
3              Wall Street Journal  
4   Bloomberg Radio and Television  
5                  Financial Times  
6                             None  
7                            Axios  
8                             None  
9                   Bloomberg News  
10                             ABC  
11                    Fox Business  
12                        Politico  
13                        CBS News  
14                     Marketplace

In [123]:
interventions_df.iloc[11,0]

'Edward. EDWARD LAWRENCE. Thanks Chair Powell questions. Edward Lawrence Fox Business. So say closer neutral rate. What percent see Committee believeswhere neutral rate CHAIR POWELL. So Ill say couple things. First write thing write Summary Economic Projections longerrun neutral rate neutral rate time supply demand balance full economys balance shocks hitting economy. That right now. So making monetary policy Fed thats question asking. So cant December 18 2024 Chair Powells Press Conference straight read longerrun numbers write think appropriate policy be. So basically given time various shocks hitting economy. And real time looking policy stance looking way hitting economy particularly look effects try move economy toward maximum employment price stability. And answer things affect economy arethat lasting permanent. And ones permanent ones would . The ones could lasting nonetheless go away time could actually affect whats sort technically appropriate neutral stance near term. So looki

In [124]:
# Rename the column 'content_cleaned' to 'questioning'
interventions_df.rename(columns={"content": "questioning"}, inplace=True)

interventions_df

questioning  \
0   Jeanna. JEANNA SMIALEK. Jeanna Smialek New Yor...   
1   Howard. HOWARD SCHNEIDER. Howard Schneider Reu...   
2   Chris. CHRIS RUGABER. Thank you. Chris Rugaber...   
3   Nick. NICK TIMIRAOS. Nick Timiraos Wall Street...   
4   Mike. MICHAEL MCKEE. Michael McKee Bloomberg R...   
5   Colby. COLBY SMITH. Thank you. Colby Smith Fin...   
6   Steve. STEVE LIESMAN. Mr. Chairman I I hear us...   
7   Neil. NEIL IRWIN. Thanks Chair Chair Powell. N...   
8   Andrew. ANDREW ACKERMAN. Hey happy holidays Mr...   
9   Amara. AMARA OMEOKWE. Thank you. Amara Omeokwe...   
10  Elizabeth. ELIZABETH SCHULZE. Thanks much Chai...   
11  Edward. EDWARD LAWRENCE. Thanks Chair Powell q...   
12  Victoria. VICTORIA GUIDA. Hi Victoria Guida Po...   
13  Kelly. KELLY OGRADY. Thanks Chair Powell. Kell...   
14  Go Nancy last question. NANCY MARSHALLGENZER. ...   

                             media  
0                   New York Times  
1                          Reuters  
2                 Associated Press  
3              Wall Street Journal  
4   Bloomberg Radio and Television  
5                  Financial Times  
6                             None  
7                            Axios  
8                             None  
9                   Bloomberg News  
10                             ABC  
11                    Fox Business  
12                        Politico  
13                        CBS News  
14                     Marketplace

In [125]:
# Function to extract full journalist names and clean the questioning column
def process_row(row):
    # Regex to match the journalist's full uppercase name
    match = re.search(r"\b([A-Z]+(?:\s[A-Z]+)*)\b", row["questioning"])

    if match:
        journalist_name = match.group(1)  # Extract the full journalist name (first and last names)
        # Remove everything before and including the first occurrence of the journalist's name
        cleaned_questioning = re.sub(rf"^.*?{re.escape(journalist_name)}\b", journalist_name, row["questioning"], count=1).strip()
    else:
        journalist_name = None
        cleaned_questioning = row["questioning"]

    return pd.Series([journalist_name, cleaned_questioning])

# Apply the function to process each row
interventions_df[["journalist_name", "questioning"]] = interventions_df.apply(process_row, axis=1)

# Reorder columns to: media, journalist_name, questioning
interventions_df = interventions_df[["media", "journalist_name", "questioning"]]

interventions_df

media       journalist_name  \
0                   New York Times        JEANNA SMIALEK   
1                          Reuters      HOWARD SCHNEIDER   
2                 Associated Press         CHRIS RUGABER   
3              Wall Street Journal         NICK TIMIRAOS   
4   Bloomberg Radio and Television         MICHAEL MCKEE   
5                  Financial Times           COLBY SMITH   
6                             None         STEVE LIESMAN   
7                            Axios            NEIL IRWIN   
8                             None       ANDREW ACKERMAN   
9                   Bloomberg News         AMARA OMEOKWE   
10                             ABC     ELIZABETH SCHULZE   
11                    Fox Business       EDWARD LAWRENCE   
12                        Politico        VICTORIA GUIDA   
13                        CBS News          KELLY OGRADY   
14                     Marketplace  NANCY MARSHALLGENZER   

                                          questioning  
0   JEANNA SMIALEK. Jeanna Smialek New York Times....  
1   HOWARD SCHNEIDER. Howard Schneider Reuters tha...  
2   CHRIS RUGABER. Thank you. Chris Rugaber Associ...  
3   NICK TIMIRAOS. Nick Timiraos Wall Street Journ...  
4   MICHAEL MCKEE. Michael McKee Bloomberg Radio T...  
5   COLBY SMITH. Thank you. Colby Smith Financial ...  
6   STEVE LIESMAN. Mr. Chairman I I hear use word ...  
7   NEIL IRWIN. Thanks Chair Chair Powell. Neil Ir...  
8   ANDREW ACKERMAN. Hey happy holidays Mr. Chairm...  
9   AMARA OMEOKWE. Thank you. Amara Omeokwe Bloomb...  
10  ELIZABETH SCHULZE. Thanks much Chair Powell. E...  
11  EDWARD LAWRENCE. Thanks Chair Powell questions...  
12  VICTORIA GUIDA. Hi Victoria Guida Politico. I ...  
13  KELLY OGRADY. Thanks Chair Powell. Kelly OGrad...  
14  NANCY MARSHALLGENZER. Hi Chair Powell. Nancy M...

In [106]:
interventions_df.iloc[12,2]

'VICTORIA GUIDA. Hi Victoria Guida Politico. I wanted make sure I understood saying beginning inflation. Are saying progress core PCE counts progress inflation even headline inflation ticks And also since thats whats projected SEP I wondering accounts Why see core PCE going potentially headline inflation ticking CHAIR POWELL. So I imagine know know goal overall headline inflation thats people experience. People dont experience core inflation experience inflation includes food energy costs. So thats overall goal. But know headline inflation contains energy food prices fluctuate reasons related tightness economy therefore really good predictors future inflation. So turns core inflation better predictor overall inflation overall inflation is. So look core inflation better measure future inflation like likely better measure December 18 2024 Chair Powells Press Conference inflation pressures exist. So thats thatsits complicated ultimately know goal headline core. So question second question

In [126]:
# Function to split questioning column into Q&A pairs
def split_questions_answers(row):
    journalist_name = row["journalist_name"]
    text = row["questioning"]

    # Initialize Q&A columns
    q_a_pairs = {}
    question_num = 1
    answer_num = 1

    # Regex pattern for splitting text
    pattern = rf"{re.escape(journalist_name)}|CHAIR POWELL"
    segments = re.split(pattern, text)

    # Iterate through segments and identify Q&A pairs
    i = 0
    while i < len(segments):
        segment = segments[i].strip()  # Remove leading and trailing spaces
        if not segment or segment == 0 or segment == ". .":  # Skip empty segments
            i += 1
            continue

        # If "CHAIR POWELL" is followed by journalist's name + "CHAIR POWELL", merge segments
        if (
            i + 1 < len(segments)
            and segments[i + 1].strip().startswith(journalist_name)
            and i + 2 < len(segments)
            and segments[i + 2].strip().startswith("CHAIR POWELL")
        ):
            # Merge current segment with the next two segments
            merged_segment = f"{segment} {segments[i + 1].strip()} {segments[i + 2].strip()}"
            q_a_pairs[f"A{answer_num}"] = merged_segment.strip()
            answer_num += 1
            i += 3  # Skip the merged segments
            continue

        # Assign segment to the correct Q/A column
        if i % 2 == 0:  # Even index = answer
            q_a_pairs[f"A{answer_num}"] = segment
            answer_num += 1
        else:  # Odd index = question
            q_a_pairs[f"Q{question_num}"] = segment
            question_num += 1

        i += 1

    return pd.Series(q_a_pairs)

# Apply the function to split questions and answers
split_df = interventions_df.apply(split_questions_answers, axis=1)

# Concatenate the original DataFrame with the split Q&A columns
result_df = pd.concat([interventions_df[["media", "journalist_name"]], split_df], axis=1)

# Reorder columns to: media, journalist_name, questioning
result_df = result_df[["media", "journalist_name", "Q1", "A1", "Q2", "A2", "Q3", "A3"]]

In [127]:
result_df

media       journalist_name  \
0                   New York Times        JEANNA SMIALEK   
1                          Reuters      HOWARD SCHNEIDER   
2                 Associated Press         CHRIS RUGABER   
3              Wall Street Journal         NICK TIMIRAOS   
4   Bloomberg Radio and Television         MICHAEL MCKEE   
5                  Financial Times           COLBY SMITH   
6                             None         STEVE LIESMAN   
7                            Axios            NEIL IRWIN   
8                             None       ANDREW ACKERMAN   
9                   Bloomberg News         AMARA OMEOKWE   
10                             ABC     ELIZABETH SCHULZE   
11                    Fox Business       EDWARD LAWRENCE   
12                        Politico        VICTORIA GUIDA   
13                        CBS News          KELLY OGRADY   
14                     Marketplace  NANCY MARSHALLGENZER   

                                                   Q1  \
0   . Jeanna Smialek New York Times. Thank taking ...   
1   . Howard Schneider Reuters thanks opportunity ...   
2   . Thank you. Chris Rugaber Associated Press. T...   
3   . Nick Timiraos Wall Street Journal. Chair Pow...   
4   . Michael McKee Bloomberg Radio Television. Ev...   
5   . Thank you. Colby Smith Financial Times. So u...   
6   . Mr. Chairman I I hear use word recalibration...   
7   . Thanks Chair Chair Powell. Neil Irwin Axios....   
8   . Hey happy holidays Mr. Chairman. Thanks taki...   
9   . Thank you. Amara Omeokwe Bloomberg News. I w...   
10  . Thanks much Chair Powell. Elizabeth Schulze ...   
11  . Thanks Chair Powell questions. Edward Lawren...   
12  . Hi Victoria Guida Politico. I wanted make su...   
13  . Thanks Chair Powell. Kelly OGrady CBS News. ...   
14  . Hi Chair Powell. Nancy MarshallGenzer Market...   

                                                   A1  \
0   . Well let let start saying wewhy cut today th...   
1   . I Id sayId Id point five six things Id start...   
2   . So I think 2018 September 2018 Tealbook alte...   
3   . Let melet find numbers. So havewe inflation ...   
4   . So risks yousorry rates talked really longer...   
5   . The unemployment rate nowis July 4.2 percent...   
6   . Were notwere renaming phaseyet. But may get ...   
7   . So look carefully financial conditions cours...   
8   . I think pretty clear weve avoided recession....   
9   . You know cant know tremendous certainty. I I...   
10  . You know breaks long answer want little bit ...   
11  . So Ill say couple things. First write thing ...   
12  . So I imagine know know goal overall headline...   
13  . Yeah inflationagain dont overreact couple mo...   
14  . No going settle that. I think know certainly...   

                                                   Q2  \
0   . Sorry one one quick followup Why I guess wou...   
1   . If I could follow You mentioned risk uncerta...   
2   . Great quick followup. Do given recent bout i...   
3   . I I guess I could follow If somebody looked ...   
4   . If I could follow asking aboutyour formulati...   
5   . If idea though additional kind softening lab...   
6   . Wonder I could follow ask much Committee loo...   
7   . So I may speaking assets havethat buoyant se...   
8   . The thing I wanted ask theyou guys noted une...   
9                                                 NaN   
10  . And look ahead next year see biggest challen...   
11  . But I think markets looking little clarity. ...   
12  . Why see core going headline potentially ticking   
13  . And one followup Lets look longterm. You pre...   
14                 . In case rule rate hike next year   

                                                   A2  \
0   . So wereto cut point I I would say way Weve r...   
1   . Actually thats thing thats happened. Whats h...   
2   . So main thing isand also point alternative s...   
3   . So thats thats thethats decision weve made a...   
4   . So confidence test raising rates know 

Hay un problema con Victoria Guida, Powell empieza a responderle pero a su vez le pregunta y ella asiente, eso quedó registrado como "onomatopeya" en la transcripcion, por tanto el A3 es continuacion del A2, y no se "recogió" el verdadero A3

In [128]:
# Initialize the Sentiment Intensity Analyzer
sia = SentimentIntensityAnalyzer()

# Function to calculate sentiment and subjectivity using VADER
def analyze_vader(text):
    if pd.isna(text) or not text.strip():  # Handle missing or empty text
        return 0.0, 0.0  # Neutral scores for empty cells
    scores = sia.polarity_scores(text)
    sentiment = scores['compound']  # Compound sentiment score (-1 to 1)
    subjectivity = scores['pos'] + scores['neg']  # Approximation for subjectivity
    return subjectivity, sentiment

# Iterate over all Q&A columns
results = []
for index, row in result_df.iterrows():  # Iterate through each row
    for column in result_df.columns:
        if column.startswith("Q") or column.startswith("A"):  # Only process Q&A columns
            subjectivity, sentiment = analyze_vader(row[column])
            results.append({
                "media": row["media"],
                "column": column,  # Q1, A1, etc.
                "text": row[column],
                "subjectivity": subjectivity,
                "sentiment": sentiment
            })

# Create a new DataFrame with the results
vader_sentiment_df = pd.DataFrame(results)

# Display the new DataFrame
print(vader_sentiment_df)

             media column                                               text  \
0   New York Times     Q1  . Jeanna Smialek New York Times. Thank taking ...   
1   New York Times     A1  . Well let let start saying wewhy cut today th...   
2   New York Times     Q2  . Sorry one one quick followup Why I guess wou...   
3   New York Times     A2  . So wereto cut point I I would say way Weve r...   
4   New York Times     Q3                                                NaN   
..             ...    ...                                                ...   
85     Marketplace     A1  . No going settle that. I think know certainly...   
86     Marketplace     Q2                 . In case rule rate hike next year   
87     Marketplace     A2  . You dont rule things completely thisin world...   
88     Marketplace     Q3                                                NaN   
89     Marketplace     A3                                                NaN   

    subjectivity  sentiment  
0        

In [109]:
sheet2 = sheets.InteractiveSheet(df=vader_sentiment_df)

https://docs.google.com/spreadsheets/d/1ZADd449Fc8PxzIpc1izUTUzldol7rQ_cm07vq_GelUg#gid=0


In [130]:
from transformers import pipeline

# Load pre-trained BERT sentiment analysis pipeline
sentiment_pipeline = pipeline("sentiment-analysis")

# Function to analyze sentiment using BERT
def analyze_bert_sentiment(text):
    result = sentiment_pipeline(text)
    sentiment = result[0]["label"]  # Sentiment label (POSITIVE, NEGATIVE, NEUTRAL)
    score = result[0]["score"]      # Confidence score
    return sentiment, score

# Create a new DataFrame with media, sentiment, and score
results = []
for media, Q1 in zip(result_df["media"], result_df["Q1"]):
    sentiment, score = analyze_bert_sentiment(Q1)
    results.append({"media": media, "sentiment": sentiment, "confidence_score": score})

bert_sentiment_df = pd.DataFrame(results)

# Display the resulting DataFrame
print(bert_sentiment_df)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cpu


                             media sentiment  confidence_score
0                   New York Times  NEGATIVE          0.992577
1                          Reuters  NEGATIVE          0.984473
2                 Associated Press  POSITIVE          0.994096
3              Wall Street Journal  NEGATIVE          0.989183
4   Bloomberg Radio and Television  NEGATIVE          0.990118
5                  Financial Times  NEGATIVE          0.978337
6                             None  POSITIVE          0.849349
7                            Axios  POSITIVE          0.999144
8                             None  POSITIVE          0.997738
9                   Bloomberg News  POSITIVE          0.992415
10                             ABC  NEGATIVE          0.981351
11                    Fox Business  NEGATIVE          0.970657
12                        Politico  NEGATIVE          0.994859
13                        CBS News  NEGATIVE          0.986294
14                     Marketplace  NEGATIVE          0

In [114]:
# Function to analyze sentiment using Stanza
# 0: NEGATIVE
# 1: NEUTRAL
# 2: POSITIVE
nlp = stanza.Pipeline('en', processors='tokenize,sentiment')
def analyze_sentiment_stanza(text):
    if pd.isna(text) or not text.strip():  # Handle missing or empty text
        return None
    doc = nlp(text)
    return [sentence.sentiment for sentence in doc.sentences]  # List of sentiment scores

# Apply the function to each Q&A column
results = []
for index, row in result_df.iterrows():
    for column in result_df.columns:
        if column.startswith("Q") or column.startswith("A"):  # Process only Q&A columns
            sentiment_scores = analyze_sentiment_stanza(row[column])
            results.append({
                "media": row["media"],
                "journalist_name": row["journalist_name"],
                "column": column,  # Q1, A1, etc.
                "text": row[column],
                "sentiment_scores": sentiment_scores
            })

# Create a new DataFrame with the sentiment analysis results
stanza_sentiment_df = pd.DataFrame(results)


INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: en (English):
| Processor | Package        |
------------------------------
| tokenize  | combined       |
| mwt       | combined       |
| sentiment | sstplus_charlm |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: sentiment
INFO:stanza:Done loading processors!


In [84]:
sheet4 = sheets.InteractiveSheet(df=stanza_sentiment_df)

https://docs.google.com/spreadsheets/d/1VCtWMTrFA-jytZthFSfNoIHsLn7Ef2TITYgYFkwXZY0#gid=0


In [115]:
# Function to calculate subjectivity based on sentiment score variance
# 0.0 --> objective
# 1.0 --> subjective

nlp = stanza.Pipeline('en', processors='tokenize,sentiment')
def calculate_subjectivity(sentiment_scores):
    if not sentiment_scores:
        return 0.0
    return sum(1 for score in sentiment_scores if score != 1) / len(sentiment_scores)  # Approximation

# Add subjectivity to the sentiment results before dropping
stanza_sentiment_df["subjectivity"] = stanza_sentiment_df["sentiment_scores"].apply(calculate_subjectivity)

# Drop the column outside the function
stanza_subjectivity_df = stanza_sentiment_df.drop(columns=["sentiment_scores"])

stanza_subjectivity_df

INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: en (English):
| Processor | Package        |
------------------------------
| tokenize  | combined       |
| mwt       | combined       |
| sentiment | sstplus_charlm |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: sentiment
INFO:stanza:Done loading processors!


media       journalist_name column  \
0   New York Times        JEANNA SMIALEK     Q1   
1   New York Times        JEANNA SMIALEK     A1   
2   New York Times        JEANNA SMIALEK     Q2   
3   New York Times        JEANNA SMIALEK     A2   
4   New York Times        JEANNA SMIALEK     Q3   
..             ...                   ...    ...   
85     Marketplace  NANCY MARSHALLGENZER     A1   
86     Marketplace  NANCY MARSHALLGENZER     Q2   
87     Marketplace  NANCY MARSHALLGENZER     A2   
88     Marketplace  NANCY MARSHALLGENZER     Q3   
89     Marketplace  NANCY MARSHALLGENZER     A3   

                                                 text  subjectivity  
0   . Jeanna Smialek New York Times. Thank taking ...      0.250000  
1   . Well let let start saying wewhy cut today th...      0.280000  
2   . Sorry one one quick followup Why I guess wou...      0.000000  
3   . So wereto cut point I I would say way Weve r...      0.333333  
4                                                 NaN      0.000000  
..                                                ...           ...  
85  . No going settle that. I think know certainly...      0.500000  
86                 . In case rule rate hike next year      1.000000  
87  . You dont rule things completely thisin world...      0.200000  
88                                                NaN      0.000000  
89                                                NaN      0.000000  

[90 rows x 5 columns]

In [116]:
sheet3 = sheets.InteractiveSheet(df=stanza_subjectivity_df)

https://docs.google.com/spreadsheets/d/1m1Z8i8B7o2c_4k2W5msFmzByeMOgJeeKLh1Yj3K24_4#gid=0


In [117]:
from gensim.models import LdaModel
from gensim.corpora import Dictionary

nlp = stanza.Pipeline(lang='en', processors='tokenize,lemma')

# Define a custom stopword list
custom_stopwords = {'i', 'know', 'so', 'that', 'over', 'and', 'you', 'the', 'a', 'it', 'to', 'for', 'in', 'on', 'do', 'say', 'go', 'be', 'see', 'not', "'s", 'thank', 'thanks', 'chair'}

# Preprocess text using Stanza
def preprocess_text_stanza(text):
    if pd.isna(text) or not text.strip():  # Handle missing or empty text
        return []
    doc = nlp(text)
    tokens = []
    for sentence in doc.sentences:
        for word in sentence.words:
            if word.text.isalnum():  # Include words and numbers
                lemma = word.lemma.lower()
                if lemma not in custom_stopwords:  # Exclude stopwords
                    tokens.append(lemma)  # Use lemma for normalization
    return tokens

# Apply preprocessing to Q&A columns
corpus = []
for column in result_df.columns:
    if column.startswith("Q") or column.startswith("A"):  # Process only Q&A columns
        preprocessed_text = result_df[column].dropna().apply(preprocess_text_stanza)
        corpus.extend(preprocessed_text)

# Create Gensim dictionary and corpus
dictionary = Dictionary(corpus)
bow_corpus = [dictionary.doc2bow(doc) for doc in corpus]

# Train LDA model
lda_model = LdaModel(bow_corpus, num_topics=5, id2word=dictionary, passes=10)

# Print the topics
for idx, topic in lda_model.print_topics(-1):
    print(f"Topic {idx}: {topic}")

INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: en (English):
| Processor | Package           |
---------------------------------
| tokenize  | combined          |
| mwt       | combined          |
| lemma     | combined_nocharlm |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: lemma
INFO:stanza:Done loading processors!


Topic 0: 0.030*"inflation" + 0.021*"year" + 0.013*"progress" + 0.013*"would" + 0.012*"cut" + 0.012*"weve" + 0.012*"market" + 0.011*"next" + 0.010*"labor" + 0.010*"think"
Topic 1: 0.033*"inflation" + 0.029*"think" + 0.016*"we" + 0.016*"good" + 0.015*"weve" + 0.014*"get" + 0.014*"month" + 0.009*"18" + 0.008*"point" + 0.008*"2024"
Topic 2: 0.016*"inflation" + 0.014*"rate" + 0.013*"service" + 0.013*"come" + 0.012*"really" + 0.012*"year" + 0.012*"market" + 0.011*"economy" + 0.011*"think" + 0.011*"good"
Topic 3: 0.020*"think" + 0.019*"inflation" + 0.013*"economy" + 0.013*"rate" + 0.012*"high" + 0.012*"thing" + 0.011*"labor" + 0.011*"year" + 0.011*"market" + 0.010*"good"
Topic 4: 0.017*"inflation" + 0.015*"look" + 0.012*"press" + 0.010*"feel" + 0.009*"powells" + 0.009*"conference" + 0.009*"december" + 0.009*"could" + 0.009*"cut" + 0.009*"2024"


In [132]:
from textblob import TextBlob


# Function to analyze sentiment and subjectivity
def analyze_text(text):
    if pd.isna(text) or not text.strip():  # Handle missing or empty text
        return 0.0, 0.0  # Neutral scores for empty cells
    blob = TextBlob(text)
    sentiment = blob.sentiment.polarity  # Sentiment score (-1.0 to 1.0)
    subjectivity = blob.sentiment.subjectivity  # Subjectivity score (0.0 to 1.0)
    return subjectivity, sentiment

# Iterate over all Q&A columns
results = []
for index, row in result_df.iterrows():  # Iterate through each row
    for column in result_df.columns:
        if column.startswith("Q") or column.startswith("A"):  # Only process Q&A columns
            subjectivity, sentiment = analyze_text(row[column])
            results.append({
                "media": row["media"],
                "column": column,  # Q1, A1, etc.
                "text": row[column],
                "subjectivity": subjectivity,
                "sentiment": sentiment
            })

# Create a new DataFrame with the results
sentiment_df = pd.DataFrame(results)

sentiment_df

media column                                               text  \
0   New York Times     Q1  . Jeanna Smialek New York Times. Thank taking ...   
1   New York Times     A1  . Well let let start saying wewhy cut today th...   
2   New York Times     Q2  . Sorry one one quick followup Why I guess wou...   
3   New York Times     A2  . So wereto cut point I I would say way Weve r...   
4   New York Times     Q3                                                NaN   
..             ...    ...                                                ...   
85     Marketplace     A1  . No going settle that. I think know certainly...   
86     Marketplace     Q2                 . In case rule rate hike next year   
87     Marketplace     A2  . You dont rule things completely thisin world...   
88     Marketplace     Q3                                                NaN   
89     Marketplace     A3                                                NaN   

    subjectivity  sentiment  
0       0.531818   0.018608  
1       0.483324   0.066194  
2       0.750000  -0.083333  
3       0.375000   0.165000  
4       0.000000   0.000000  
..           ...        ...  
85      0.370238   0.252381  
86      0.000000   0.000000  
87      0.566667   0.246667  
88      0.000000   0.000000  
89      0.000000   0.000000  

[90 rows x 5 columns]

La mayoria de los periodistas suele preguntar la opinion de Powell respecto a las condiciones economicas, por lo que estas se toman con un alto grado de subjetividad; Powell por su parte suele enunciar hechos antes de emitir su opinion por lo que las respuestas estarán en un termino intermedio en lo que refiere a objetividad/subjetividad.

La mayoría de las preguntas de los periodistas suelen tener un tono neutro, sin embargo, unos dan un enfoque negativo cuando emiten su pregunta. Powell, se muestra casi con una proporcion de 50/50 entre lo neutro y lo positivo

Topic 0: 0.030*"inflation" + 0.021*"year" + 0.013*"progress" + 0.013*"would" + 0.012*"cut" + 0.012*"weve" + 0.012*"market" + 0.011*"next" + 0.010*"labor" + 0.010*"think"

Topic 1: 0.033*"inflation" + 0.029*"think" + 0.016*"we" + 0.016*"good" + 0.015*"weve" + 0.014*"get" + 0.014*"month" + 0.009*"18" + 0.008*"point" + 0.008*"2024"

Topic 2: 0.016*"inflation" + 0.014*"rate" + 0.013*"service" + 0.013*"come" + 0.012*"really" + 0.012*"year" + 0.012*"market" + 0.011*"economy" + 0.011*"think" + 0.011*"good"

Topic 3: 0.020*"think" + 0.019*"inflation" + 0.013*"economy" + 0.013*"rate" + 0.012*"high" + 0.012*"thing" + 0.011*"labor" + 0.011*"year" + 0.011*"market" + 0.010*"good"

Topic 4: 0.017*"inflation" + 0.015*"look" + 0.012*"press" + 0.010*"feel" + 0.009*"powells" + 0.009*"conference" + 0.009*"december" + 0.009*"could" + 0.009*"cut" + 0.009*"2024"

El tema en el Q&A se centra fundamentalmente en la inflación y en la opinión de Powell

#Analisis Powell remarks

In [58]:
# Convert the DataFrame column to a list to create a corpus
corpus_PR = powell_remarks_df["content"].dropna().tolist()
corpus_PR

['December 18, 2024 Chair Powell’s Press Conference \nTranscript of Chair Powell’s Press Conference\nDecember 18, 2024\nCHAIR POWELL. Good afternoon. My colleagues and I remain squarely focused on\nachieving our dual-mandate goals of maximum employment and stable prices for the benefit of\nthe American people. The economy is strong overall and has made significant progress toward\nour, our goals over the past two years. The labor market has cooled from its formerly overheated\nstate and remains solid. Inflation has moved much closer to our 2 percent longer-run goal.\nWe’re committed to maintaining our economy’s strength by supporting maximum\nemployment and returning inflation to our 2 percent goal. To that end, today, the Federal Open\nMarket Committee decided to take another step in reducing the degree of policy restraint by\nlowering our policy interest rate by ¼ percentage point. We also decided to continue to reduce\nour securities holdings. I’ll have more to say about monetary po

In [59]:
# Define patterns to remove unwanted elements
date_pattern = r"\b(January|February|March|April|May|June|July|August|September|October|November|December) \d{1,2}, \d{4}\b"  # Matches date format
title_pattern = r"Chair Powell’s Press Conference"  # Matches title
version_pattern = r"\bFINAL\b"  # Matches transcript version
page_counter_pattern = r"Page \d+ of \d+"  # Matches page counter
transcript_pattern = r"Transcript of Chair Powell’s Press Conference"  # Matches transcript title

# Combine all patterns into one regex
unwanted_pattern = re.compile(
    rf"({date_pattern})|({title_pattern})|({version_pattern})|({page_counter_pattern})|({transcript_pattern})")

# Clean the text by removing unwanted elements
# corpus_PR is a list, so you need to iterate over its elements
corpus_clean = []
for text in corpus_PR:  # Iterate through each element in corpus_PR
    cleaned_text = re.sub(unwanted_pattern, "", text) # Apply re.sub to each string element
    corpus_clean.append(cleaned_text)  # Add the cleaned text to the new list

# Remove extra spaces created by the replacements
corpus_clean = [re.sub(r"\s+", " ", text).strip() for text in corpus_clean]

# If you want a single string as output, join the elements of corpus_clean
corpus_clean = " ".join(corpus_clean)

corpus_clean

'CHAIR POWELL. Good afternoon. My colleagues and I remain squarely focused on achieving our dual-mandate goals of maximum employment and stable prices for the benefit of the American people. The economy is strong overall and has made significant progress toward our, our goals over the past two years. The labor market has cooled from its formerly overheated state and remains solid. Inflation has moved much closer to our 2 percent longer-run goal. We’re committed to maintaining our economy’s strength by supporting maximum employment and returning inflation to our 2 percent goal. To that end, today, the Federal Open Market Committee decided to take another step in reducing the degree of policy restraint by lowering our policy interest rate by ¼ percentage point. We also decided to continue to reduce our securities holdings. I’ll have more to say about monetary policy after briefly reviewing economic developments. Recent indicators suggest that economic activity has continued to expand at 

In [60]:
# Function to clean text in Powell_remarks
def clean_text(text):
    # Remove line breaks and replace them with spaces
    text = text.replace("\n", " ")

    # Generalized regex for human sounds and pauses
    human_sounds_regex = r"\b(?:[Mm]+[Hh]\s[Hh]|[Uu]+[Mm]|[Hh]+[Mm]+|[Aa]+[Hh]|[Oo]+[Hh]|[Ee]+[Rr]*|[Uu]+[Hh]+|[Mm]+|[Hh]+|[Uu]+|[Pp]+[Hh]+|[Er]+)\b"

    # Remove human sounds using the regex
    text = re.sub(human_sounds_regex, "", text)

    # Fix stuttering (e.g., "s-sample" -> "sample")
    text = re.sub(r"\b([a-zA-Z])-\1\b", r"\1", text)

    # Eliminar prefijos 'b' si existen y no están seguidos de ninguna letra
    text = re.sub(r'\b(b)\b', '', text)

    # Keep numbers and periods while removing other special characters
    text = re.sub(r"[^a-zA-Z0-9.\sáéíóúüñçÁÉÍÓÚÜÑÇ]", '', text)

    # Replace NaN values with an empty string
    if pd.isnull(text):
        text = ""

    # Eliminar palabras vacías (stopwords)
    stop_words = set(stopwords.words('english'))  # Usar la lista de stopwords en inglés
    text = ' '.join(word for word in text.split() if word not in stop_words)

    # Remove all hyphens and replace them with a single space
    text = re.sub(r"-", " ", text)

    # Remove all apostrophes
    text = re.sub(r"'", "", text)

    # Replace multiple spaces with a single space
    text = re.sub(r"\s+", " ", text).strip()

    return text

# corpus_clean is a string, so you need to call clean_text directly
corpus_filtered = clean_text(corpus_clean)

corpus_filtered

'CHAIR POWELL. Good afternoon. My colleagues I remain squarely focused achieving dualmandate goals maximum employment stable prices benefit American people. The economy strong overall made significant progress toward goals past two years. The labor market cooled formerly overheated state remains solid. Inflation moved much closer 2 percent longerrun goal. Were committed maintaining economys strength supporting maximum employment returning inflation 2 percent goal. To end today Federal Open Market Committee decided take another step reducing degree policy restraint lowering policy interest rate percentage point. We also decided continue reduce securities holdings. Ill say monetary policy briefly reviewing economic developments. Recent indicators suggest economic activity continued expand aat solid pace. GDP rose annual rate 2.8 percent third quarter pace second quarter. Growth consumer spending remained resilient investment equipment intangibles strengthened. In contrast activity housin

In [66]:
nlp = stanza.Pipeline(lang='en', processors='tokenize,sentiment')

# Function to clean and ensure proper formatting for Stanza
def prepare_corpus_for_sentiment(corpus):
    # Ensure periods are properly used to separate sentences
    cleaned_corpus = corpus.replace("\n", " ")  # Replace newlines with spaces
    cleaned_corpus = cleaned_corpus.strip()  # Remove leading/trailing spaces
    return cleaned_corpus

# Clean the corpus
cleaned_corpus = prepare_corpus_for_sentiment(corpus_filtered)

# Process the cleaned corpus with Stanza
doc = nlp(cleaned_corpus)

# Extract sentiment for each sentence
results = []
for idx, sentence in enumerate(doc.sentences):
    results.append({
        "sentence": sentence.text,  # Extract the sentence
        "sentiment": sentence.sentiment  # Extract the sentiment score
    })

# Display results
for result in results:
    print(f"Sentence: {result['sentence']}")
    print(f"Sentiment: {result['sentiment']} (0: Negative, 1: Neutral, 2: Positive)\n")

INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: en (English):
| Processor | Package        |
------------------------------
| tokenize  | combined       |
| mwt       | combined       |
| sentiment | sstplus_charlm |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: sentiment
INFO:stanza:Done loading processors!


Sentence: CHAIR POWELL.
Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

Sentence: Good afternoon.
Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

Sentence: My colleagues I remain squarely focused achieving dualmandate goals maximum employment stable prices benefit American people.
Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

Sentence: The economy strong overall made significant progress toward goals past two years.
Sentiment: 2 (0: Negative, 1: Neutral, 2: Positive)

Sentence: The labor market cooled formerly overheated state remains solid.
Sentiment: 2 (0: Negative, 1: Neutral, 2: Positive)

Sentence: Inflation moved much closer 2 percent longerrun goal.
Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

Sentence: Were committed maintaining economys strength supporting maximum employment returning inflation 2 percent goal.
Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

Sentence: To end today Federal Open Market Committee decided take another step reducing deg

In [69]:
import statistics

# Extract sentiment scores from the results
sentiment_scores = [result["sentiment"] for result in results]

# Calculate mean and mode
mean_sentiment = statistics.mean(sentiment_scores)
mode_sentiment = statistics.mode(sentiment_scores)

# Display results
print(f"Mean Sentiment: {mean_sentiment}")
print(f"Mode Sentiment: {mode_sentiment}")

Mean Sentiment: 1.0357142857142858
Mode Sentiment: 1


In [75]:
# Function to analyze subjectivity using TextBlob
def analyze_subjectivity_textblob(corpus):
    blob = TextBlob(corpus)
    results = []
    for sentence in blob.sentences:
        results.append({
            "sentence": str(sentence),
            "subjectivity": sentence.subjectivity  # Subjectivity score (0: Objective, 1: Subjective)
        })
    return results

# Analyze the "cleaned_corpus"
subjectivity_results_textblob = analyze_subjectivity_textblob(cleaned_corpus)

# Display the results
for result in subjectivity_results_textblob:
    print(f"Sentence: {result['sentence']}")
    print(f"Subjectivity: {result['subjectivity']}\n")

Sentence: CHAIR POWELL.
Subjectivity: 0.0

Sentence: Good afternoon.
Subjectivity: 0.6000000000000001

Sentence: My colleagues I remain squarely focused achieving dualmandate goals maximum employment stable prices benefit American people.
Subjectivity: 0.0

Sentence: The economy strong overall made significant progress toward goals past two years.
Subjectivity: 0.46458333333333335

Sentence: The labor market cooled formerly overheated state remains solid.
Subjectivity: 0.05

Sentence: Inflation moved much closer 2 percent longerrun goal.
Subjectivity: 0.2

Sentence: Were committed maintaining economys strength supporting maximum employment returning inflation 2 percent goal.
Subjectivity: 0.25

Sentence: To end today Federal Open Market Committee decided take another step reducing degree policy restraint lowering policy interest rate percentage point.
Subjectivity: 0.5

Sentence: We also decided continue reduce securities holdings.
Subjectivity: 0.0

Sentence: Ill say monetary policy b

In [74]:
from nltk.sentiment import SentimentIntensityAnalyzer

# Initialize VADER Sentiment Intensity Analyzer
sia = SentimentIntensityAnalyzer()

# Function to calculate subjectivity as non-neutral sentiment ratio
def analyze_subjectivity_vader(corpus):
    sentences = corpus.split(". ")  # Split corpus into sentences
    results = []
    for sentence in sentences:
        if sentence.strip():  # Skip empty sentences
            scores = sia.polarity_scores(sentence)
            subjectivity = scores["pos"] + scores["neg"]  # Approximation for subjectivity
            results.append({
                "sentence": sentence,
                "subjectivity": subjectivity
            })
    return results

# Analyze the "cleaned_corpus"
subjectivity_results_vader = analyze_subjectivity_vader(cleaned_corpus)

# Display the results
for result in subjectivity_results_vader:
    print(f"Sentence: {result['sentence']}")
    print(f"Subjectivity: {result['subjectivity']}\n")

Sentence: CHAIR POWELL
Subjectivity: 0.0

Sentence: Good afternoon
Subjectivity: 0.744

Sentence: My colleagues I remain squarely focused achieving dualmandate goals maximum employment stable prices benefit American people
Subjectivity: 0.394

Sentence: The economy strong overall made significant progress toward goals past two years
Subjectivity: 0.467

Sentence: The labor market cooled formerly overheated state remains solid
Subjectivity: 0.167

Sentence: Inflation moved much closer 2 percent longerrun goal
Subjectivity: 0.0

Sentence: Were committed maintaining economys strength supporting maximum employment returning inflation 2 percent goal
Subjectivity: 0.477

Sentence: To end today Federal Open Market Committee decided take another step reducing degree policy restraint lowering policy interest rate percentage point
Subjectivity: 0.20800000000000002

Sentence: We also decided continue reduce securities holdings
Subjectivity: 0.268

Sentence: Ill say monetary policy briefly reviewi

In [86]:
# Create a list of dictionaries with subjectivity scores and sentences
subjectivity_data = []
for vader_result, textblob_result in zip(subjectivity_results_vader, subjectivity_results_textblob):
    subjectivity_data.append({
        "vader_subjectivity": vader_result["subjectivity"],
        "textblob_subjectivity": textblob_result["subjectivity"],
        "sentence": vader_result["sentence"]
    })

# Create the DataFrame from the list of dictionaries
subjectivity_df = pd.DataFrame(subjectivity_data)

subjectivity_df

vader_subjectivity  textblob_subjectivity  \
0                0.000               0.000000   
1                0.744               0.600000   
2                0.394               0.000000   
3                0.467               0.464583   
4                0.167               0.050000   
5                0.000               0.200000   
6                0.477               0.250000   
7                0.208               0.500000   
8                0.268               0.000000   
9                0.286               0.511111   
10               0.328               0.183333   
11               0.000               0.316667   
12               0.435               0.000000   
13               0.367               0.625000   
14               0.627               0.366667   
15               0.000               0.250000   
16               0.219               0.200000   
17               0.242               0.100000   
18               0.167               0.375000   
19               0.357               0.400000   
20               0.407               0.250000   
21               0.000               0.166220   
22               0.451               0.875000   
23               0.182               0.000000   
24               0.155               0.125000   
25               0.072               0.883333   
26               0.208               0.156250   
27               0.000               0.416667   
28               0.000               0.100000   
29               0.286               0.000000   
30               0.276               0.650000   
31               0.136               0.000000   
32               0.370               0.000000   
33               0.097               0.238889   
34               0.189               0.000000   
35               0.219               0.400000   
36               0.174               0.350000   
37               0.104               0.000000   
38               0.000               0.000000   
39               0.000               0.525000   
40               0.000               0.250000   
41               0.000               0.375000   
42               0.000               0.000000   
43               0.405               0.300000   
44               0.191               0.377778   
45               0.346               0.666667   
46               0.452               0.000000   
47               0.073               0.300000   
48               0.362               0.500000   
49               0.180               0.000000   
50               0.336               0.250000   
51               0.545               0.000000   
52               0.000               0.000000   
53               0.000               0.066667   
54               0.000               0.000000   
55               0.455               0.000000   

                                             sentence  
0                                        CHAIR POWELL  
1                                      Good afternoon  
2   My colleagues I remain squarely focused achiev...  
3   The economy strong overall made significant pr...  
4   The labor market cooled formerly overheated st...  
5   Inflation moved much closer 2 percent longerru...  
6   Were committed maintaining economys strength s...  
7   To end today Federal Open Market Committee dec...  
8   We also decided continue reduce securities hol...  
9   Ill say monetary policy briefly reviewing econ...  
10  Recent indicators suggest economic activity co...  
11  GDP rose annual rate 2.8 percent third quarter...  
12  Growth consumer spending remained resilient in...  
13           In contrast activity housing sector weak  
14  Overall improving supply conditions supported ...  
15                                  economy past year  
16  In Summary Economic Projections Committee part...  
17            In labor market conditions remain solid  
18  Payroll job gains slowed earlier year averagin...  
19  The unemployment rate higher year ago 4.2 perc...  
20  Nominal wage growth eased past

In [87]:
# Calculate the mean for each subjectivity
mean_vader_subjectivity = subjectivity_df["vader_subjectivity"].mean()
mean_textblob_subjectivity = subjectivity_df["textblob_subjectivity"].mean()

# Display the means
print(f"Mean VADER Subjectivity: {mean_vader_subjectivity}")
print(f"Mean TextBlob Subjectivity: {mean_textblob_subjectivity}")

Mean VADER Subjectivity: 0.22185714285714284
Mean TextBlob Subjectivity: 0.2431219883786848


In [89]:
nlp = stanza.Pipeline(lang='en', processors='tokenize,lemma')

# Define a custom stopword list
custom_stopwords = {'i', 'know', 'so', 'that', 'over', 'and', 'you', 'the', 'a', 'it', 'to', 'for', 'in', 'on', 'do', 'say', 'go', 'be', 'see', 'not', "'s", 'thank', 'thanks', 'chair', 'we'}

# Preprocess text using Stanza
def preprocess_text_stanza(text):
    if not text.strip():  # Handle empty text
        return []
    doc = nlp(text)
    tokens = []
    for sentence in doc.sentences:
        for word in sentence.words:
            if word.text.isalnum():  # Include words and numbers
                lemma = word.lemma.lower()
                if lemma not in custom_stopwords:  # Exclude stopwords
                    tokens.append(lemma)  # Use lemma for normalization
    return tokens

# Preprocess the cleaned_corpus
processed_corpus = [preprocess_text_stanza(sentence) for sentence in cleaned_corpus.split('.') if sentence.strip()]

# Create Gensim dictionary and Bag-of-Words corpus
dictionary = Dictionary(processed_corpus)
bow_corpus = [dictionary.doc2bow(doc) for doc in processed_corpus]

# Train the LDA model
lda_model = LdaModel(bow_corpus, num_topics=5, id2word=dictionary, passes=10)

# Print the topics
for idx, topic in lda_model.print_topics(-1):
    print(f"Topic {idx}: {topic}")

INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: en (English):
| Processor | Package           |
---------------------------------
| tokenize  | combined          |
| mwt       | combined          |
| lemma     | combined_nocharlm |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: lemma
INFO:stanza:Done loading processors!


Topic 0: 0.045*"goal" + 0.033*"inflation" + 0.029*"2" + 0.026*"percent" + 0.026*"employment" + 0.023*"remain" + 0.023*"economy" + 0.021*"maximum" + 0.018*"support" + 0.018*"strong"
Topic 1: 0.028*"rate" + 0.019*"adjustment" + 0.019*"policy" + 0.019*"evolve" + 0.019*"consider" + 0.011*"economy" + 0.010*"good" + 0.010*"monetary" + 0.010*"solid" + 0.010*"activity"
Topic 2: 0.020*"policy" + 0.020*"market" + 0.020*"labor" + 0.020*"progress" + 0.020*"year" + 0.020*"3" + 0.017*"percent" + 0.014*"next" + 0.014*"much" + 0.013*"federal"
Topic 3: 0.038*"percent" + 0.034*"year" + 0.023*"2" + 0.021*"rate" + 0.019*"remain" + 0.019*"4" + 0.014*"participant" + 0.014*"range" + 0.013*"past" + 0.011*"fund"
Topic 4: 0.038*"policy" + 0.021*"projection" + 0.021*"price" + 0.017*"market" + 0.017*"2" + 0.013*"median" + 0.013*"labor" + 0.013*"inflation" + 0.013*"pce" + 0.012*"percent"


#Intervencion de Powell

El análisis de la intervencion de Powell revela que su discurso es en su mayoria neutral, lo cual está en concordancia con la "naturaleza" del evento; pues consiste en informar sobre el estado actual de la economia de Estados Unidos, cómo influyeron las decisiones respecto a la subida, bajada o mantenimiento de la tasas de interés de meses anteriores en los indicadores económicos. Tambien se refiere a las proyecciones existentes de inflación y en qué etapa del proceso para alcanzarla se encuentran.

**Sentence**: Our success delivering goals matters Americans.

Sentiment: 2 (0: Negative, 1: Neutral, 2: Positive)




**Sentence**: We understand actions affect communities families businesses across country.

Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)


**Sentence**: The median projection SEP total PCE inflation 2.4 percent year 2.5 percent next year somewhat higher projected September.

Sentiment: 2 (0: Negative, 1: Neutral, 2: Positive)

**Sentence**: Thereafter median projection falls 2 percent objective.

Sentiment: 0 (0: Negative, 1: Neutral, 2: Positive)




**Sentence**: Our monetary policy actions guided dual mandate promote maximum employment stable prices American people.

Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

**Sentence**: Growth consumer spending remained resilient investment equipment intangibles strengthened.

Sentiment: 2 (0: Negative, 1: Neutral, 2: Positive)

**Sentence**: In contrast activity housing sector weak.

Sentiment: 0 (0: Negative, 1: Neutral, 2: Positive)

**Sentence**: With todays action lowered policy rate full percentage point peak policy stance significantly less restrictive.

Sentiment: 0 (0: Negative, 1: Neutral, 2: Positive)

**Sentence**: We therefore cautious consider adjustments policy rate.

Sentiment: 1 (0: Negative, 1: Neutral, 2: Positive)

En los ejemplos anteriores tambien se aprecia que, a interpretacion de un lector, en algunas oraciones el sentimiento positivo o negativo está más "marcado que en otras", sin embargo, en la mayoria de oraciones se percibe neutralidad

**vader:**0.545 	**textBlob:**0.0	Our success delivering goals matters Americans

**vader:**0.0     **textBlob:**	0.0	We understand actions affect communities families businesses across country


**vader:**0.0	    **textBlob:**0.417	The median projection SEP total PCE inflation 2.4 percent year 2.5 percent next year somewhat higher projected September


**vader:**0.0	    **textBlob:**0.1	Thereafter median projection falls 2 percent objective

**vader:**0.097  	**textBlob:**0.24	With todays action lowered policy rate full percentage point peak policy stance significantly less restrictive


**vader:**0.189  	**textBlob:**0.0	We therefore cautious consider adjustments policy rate

Respecto a la subjetividad, se supone que la mayoria del texto se perciba como objativo, pues se está hablando con datos y hechos, no con suposiciones; esto se puede interpretar de los valores medios de la subjetividad por los modelos empleados:

Mean VADER Subjectivity: 0.22185714285714284

Mean TextBlob Subjectivity: 0.2431219883786848

El modelo VADER, que basa su analisis de subjetividad en el analisis de sentimiento, acierta con "cierta frecuencia" en lo que un lector podría definir como objetivo, pues se asume que los textos interpretados como "neutrales" sean los de menos score de subjetividad (los más objetivos)

En lo que respecta el tema y palabras frecuentes:


Topic 0: 0.045*"goal" + 0.033*"inflation" + 0.029*"2" + 0.026*"percent" + 0.026*"employment" + 0.023*"remain" + 0.023*"economy" + 0.021*"maximum" + 0.018*"support" + 0.018*"strong"

Topic 1: 0.028*"rate" + 0.019*"adjustment" + 0.019*"policy" + 0.019*"evolve" + 0.019*"consider" + 0.011*"economy" + 0.010*"good" + 0.010*"monetary" + 0.010*"solid" + 0.010*"activity"

Topic 2: 0.020*"policy" + 0.020*"market" + 0.020*"labor" + 0.020*"progress" + 0.020*"year" + 0.020*"3" + 0.017*"percent" + 0.014*"next" + 0.014*"much" + 0.013*"federal"

Topic 3: 0.038*"percent" + 0.034*"year" + 0.023*"2" + 0.021*"rate" + 0.019*"remain" + 0.019*"4" + 0.014*"participant" + 0.014*"range" + 0.013*"past" + 0.011*"fund"

Topic 4: 0.038*"policy" + 0.021*"projection" + 0.021*"price" + 0.017*"market" + 0.017*"2" + 0.013*"median" + 0.013*"labor" + 0.013*"inflation" + 0.013*"pce" + 0.012*"percent"


Frases como "labor market", "federal funds rate", "maximum employement", "monetary policy"; han sido bastante frecuentes, asi como "inflation", pues Powell en aras de tomar una decision respecto a la "politica monetaria" de Estados Unidos, en su poder está decidir las "tasas de interés" con el objetivo de tratar de controlar la "inflación"; y para ello debe basar su decisión en las condiciones del "mercado laboral" y la "tasa de desempleo"

En el proyecto quedaria por refinar el tratamiento a los datos y ampliar el analisis en especifico a medios